rust/windows-client/src-tauri/src/client/crash_handling.rs

```diff
@@ -0,0 +1,162 @@
+    let exe = std::env::current_exe().context("unable to find our own exe path")?;
+    let mut server = None;
+
+    for _ in 0..10 {
+        // Create the crash client first so we can error out if another instance of
+        // the Firezone client is already using this socket for crash handling.
+        if let Ok(client) = Client::with_name(SOCKET_NAME) {
+            return Ok((
+                client,
+                server.ok_or_else(|| {
+                    anyhow!(
+                        "should be impossible to make a client if we didn't make the server yet"
+                    )
+                })?,
+            ));
+        }
+
+        server = Some(
+            std::process::Command::new(&exe)
+                .arg("crash-handler-server")
+                .spawn()
+                .context("unable to spawn server process")?,
+        );
+
+        // Give it time to start
+        std::thread::sleep(std::time::Duration::from_millis(100));
+    }
+
+    bail!("Couldn't set up crash handler server")
+}
+
+/// Main function for the server process, for out-of-process crash handling.
+///
+/// The server process seems to be the preferred method,
+/// since it's hard to run complex code in a process
+/// that's already crashed and likely suffered memory corruption.
+///
+/// <https://jake-shadle.github.io/crash-reporting/#implementation>
+/// <https://chromium.googlesource.com/breakpad/breakpad/+/master/docs/getting_started_with_breakpad.md#terminology>
+pub(crate) fn server() -> anyhow::Result<()> {
+    let mut server = Server::with_name(SOCKET_NAME)?;
+
+    let ab = std::sync::atomic::AtomicBool::new(false);
+
+    struct Handler;
+
+    impl minidumper::ServerHandler for Handler {
+        /// Called when a crash has been received and a backing file needs to be
+        /// created to store it.
+        fn create_minidump_file(&self) -> Result<(File, PathBuf), std::io::Error> {
+            let dump_path = get_known_folder_path(KnownFolder::ProgramData)
+                .expect("should be able to find C:/ProgramData")
+                .join(crate::client::gui::TAURI_ID)
+                .join("dumps")
+                .join("last_crash.dmp");
+
+            if let Some(dir) = dump_path.parent() {
+                if !dir.try_exists()? {
+                    std::fs::create_dir_all(dir)?;
+                }
+            }
+            let file = File::create(&dump_path)?;
+            Ok((file, dump_path))
+        }
+
+        /// Called when a crash has been fully written as a minidump to the provided
+        /// file. Also returns the full heap buffer as well.
+        fn on_minidump_created(
+            &self,
+            result: Result<minidumper::MinidumpBinary, minidumper::Error>,
+        ) -> minidumper::LoopAction {
+            match result {
+                Ok(mut md_bin) => {
+                    let _ = md_bin.file.flush();
+                    tracing::info!("wrote minidump to disk");
```

Comment:
This happens in the crash handler process, so unfortunately these tracing logs don't actually go anywhere useful I think.

It might be hard for the crash handler process to append to the same connlib log file that the main process is using, too.

I was thinking of having the main process detect non-graceful shutdowns so even if the crash handler doesn't trigger it can have a warning saying "Hey I didn't shut down gracefully last time, maybe it's a crash or maybe Task Manager killed me, check this path if you want"

Reply:
Are these `tracing` calls going anywhere? I don't see any subscriber.

Reply:
They don't go anywhere. I assume it won't work to have two separate processes or subscribers writing to the same log files, but maybe I could set up a separate log dir for the crash handler process?